Better handle avatars

rullzer opened the pull request:
Fixes #7497
Fixes #7500

- [X] Do not use placeholder code to avoid flashing of the image
- [x] Actually delete generated avatar on displayname change
- [x] Refresh avatar after displayname change in settings
- [x] Show loading icon
- [x] Allow deletion of profile picture https://github.com/nextcloud/server/issues/7500

Future work:
- [ ] Not all letters are properly centerered

rullzer committed:
Since we now always generate an avatar do not load the placeholder

The js and php code differ ever so slightly. So having the placeholder
for a second and then the image is just weird.

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Remove generated avatar on displayname change

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Update avatar on displayname change

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

danxuliu commented:
>  Do not use placeholder code to avoid flashing of the image

Before, if a display name was given when calling `avatar()`, a placeholder was shown until the avatar image was loaded; now nothing is shown until the avatar image is loaded. Maybe it would be better to replace the placeholder with a loading icon instead? The same approach could be used too when calling `avatar()` without giving a display name (even before this change nothing was shown in that case until the image loaded).

Also, if the problem is the flash caused when the placeholder is replaced by the image, another option could be to show the placeholder and then, when the image is loaded, fade out the placeholder and fade in the image.

Or maybe both, I mean, show the loading icon instead of the placeholder and also fade out the loading icon while fading in the avatar image once it is loaded (although as the animation should be pretty short it would end causing a flash anyway...).

In any case, with the current code if a display name is given and the avatar image fails to load no avatar is shown at all, but [if no display name is given and the avatar image fails to load then a placeholder is shown](https://github.com/nextcloud/server/blob/f7e6da9a8eac68cb0aa0fc9a593d3fb05e41fb89/core/js/jquery.avatar.js#L112); at the very least the behaviour should be the same in both cases.

@nextcloud/designers

rullzer commented:
@danxuliu I'm afraid my JS and design skills are lacking here. So feel free to hack in a loading animation :)

skjnldsv committed:
Add loading icon on avatar

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv commented:
> @danxuliu I'm afraid my JS and design skills are lacking here. So feel free to hack in a loading animation :)
Done

rullzer committed:
Still call callback once avatar is loaded

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer commented:
Ok let get this in before it becomes to much of a mess :)

MorrisJobke commented on lib/public/IAvatar.php:
```diff
@@ -77,4 +77,10 @@ public function remove();
 	public function getFile($size);
+
+	/**
+	 * Is this a generated avatar
```
It looks like the comment is wrong 😉

MorrisJobke commented on core/js/jquery.avatar.js:
```diff
@@ -136,21 +136,21 @@
-				$div.show();
 				$div.text('');
 				$div.append(img);
 				$div.clearimageplaceholder();
```
I guess this is then also not needed anymore.

skjnldsv replied:
Yes it is because the div is cleared with the loading class as well
I think it could also interfer with the uploading of a new avatar.

MorrisJobke reviewed (dismissed):
Code makes sense, tested and works 👍

rullzer commented:
@skjnldsv you forgot to sign off ;)

skjnldsv commented:
Rebase and ready to 🚢

danxuliu reviewed (dismissed):
The loading icon should be shown too when `avatar()` is called without a display name.

If a display name is given and the avatar image fails to load no avatar is shown at all, but [if no display name is given and the avatar image fails to load then a placeholder is shown](https://github.com/nextcloud/server/blob/f7e6da9a8eac68cb0aa0fc9a593d3fb05e41fb89/core/js/jquery.avatar.js#L112); the behaviour should be the same in both cases.

In the user settings the button to delete the avatar image is shown even if the image was auto-generated, and removing the auto-generated image just recreates it again, so it looks like nothing happened when clicking the button (except for a loading icon being shown). However I do not know if we have any way to distinguish between a user uploaded image and an auto-generated image in the avatars... @rullzer

rullzer commented:
> In the user settings the button to delete the avatar image is shown even if the image was auto-generated, and removing the auto-generated image just recreates it again, so it looks like nothing happened when clicking the button (except for a loading icon being shown). However I do not know if we have any way to distinguish between a user uploaded image and an auto-generated image in the avatars... @rullzer

Currently not. While not ideal I'd say leave it for now and we look for a proper solution.

danxuliu commented:
@rullzer 
>  Currently not. While not ideal I'd say leave it for now and we look for a proper solution.

I agree.

@skjnldsv Do you want to fix the other two issues I mentioned or do you prefer me to do it?

By the way, I intended to add JavaScript unit tests to ensure that the callback was called when a display name was given, but although I tried hard I have not been able to spy on the [Image constructor](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/Image) (to then call `load` on the `Image` instance created inside `avatar()` and check that it called the callback); honestly I do not understand why it does not work, so if someone knows how to solve that... :-)

skjnldsv commented:
@danxuliu I'm a bit busy right now, I wouldn't mind if you could take care of it :)
You're awesome! 😉

juliusknorr committed:
Fix fallback when image loading fails

With the new avatar endpoint there is no difference between unknown
users and errors when generating the placeholder avatar. Therefore the
avatar function will now show the old placeholder if both a user and
displayname was given as parameters.

In case there is no displayname provided we cannot build the proper
placeholder so the unknown user placeholder is shown.

The displayname is not required for the avatar anymore, so we can
get rid of the old code path for placeholders.

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

juliusknorr committed:
Hide delete button if the avatar is a generated one

Signed-off-by: Julius Härtl <jus@bitgrid.net>

juliushaertl commented:
I pushed two commits to fix the issues described by @danxuliu. I reduced the complexity of the avatar function quite a bit. We don't rely on the displayname for avatar placeholders anymore, therefore we can totally get rid of the old code path for placeholder generation.

Also for setups without gd installed i keep generating the placeholder in case there is a displayname provided to the avatar plugin.

danxuliu commented:
>  I pushed two commits to fix the issues described by @danxuliu.

Nice, thanks! :-D

But... unsigned commit (first one) and failing tests :-)

And another detail: the callback should be called too when an error happens (like it was done in the old code path).

Also I am wondering if we should write a log to the console or maybe a warning when `ie8fix` and `hidedefault` parameters are given to make app developers aware that those parameter are no longer used and that they will be removed for Nextcloud 14.

rullzer commented:
> Also I am wondering if we should write a log to the console or maybe a warning when ie8fix and hidedefault parameters are given to make app developers aware that those parameter are no longer used and that they will be removed for Nextcloud 14.

Sure. But this will have to go into a list of breaking changes as it forces app that use this to move to a NC14 only version. But fine by me.

rullzer committed:
Also do callback on the failing function

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer commented:
Failing tests are due to image mocking of the image function magic... lets see if we can somehow solve that properly :S

rullzer committed:
Fix tests

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer commented:
Fixed the tests I think

juliusknorr committed:
Fix failing AvatarTest

Signed-off-by: Julius Härtl <jus@bitgrid.net>

rullzer commented:
@danxuliu I think your comments are addressed. Can you verify?

MorrisJobke reviewed (approved):
Tested and works 👍

MorrisJobke commented:
@danxuliu @skjnldsv @blizzz Please review
